src/query/src/parser.rs

```diff
@@ -54,25 +67,83 @@ impl QueryLanguageParser {
     }
 
     // TODO(ruihang): implement this method when parser is ready.
-    pub fn parse_promql(promql: &str) -> Result<QueryStatement> {
+    pub fn parse_promql(query: &PromQuery) -> Result<QueryStatement> {
         let _timer = timer!(METRIC_PARSE_PROMQL_ELAPSED);
 
-        let prom_expr = promql_parser::parser::parse(promql)
+        let expr = promql_parser::parser::parse(&query.query)
             .map_err(|msg| BoxedError::new(PlainError::new(msg, StatusCode::InvalidArguments)))
-            .context(QueryParseSnafu { query: promql })?;
+            .context(QueryParseSnafu {
+                query: &query.query,
+            })?;
+
+        let start = Self::parse_promql_timestamp(&query.start)
+            .map_err(BoxedError::new)
+            .context(QueryParseSnafu {
+                query: &query.query,
+            })?;
+
+        let end = Self::parse_promql_timestamp(&query.end)
+            .map_err(BoxedError::new)
+            .context(QueryParseSnafu {
+                query: &query.query,
+            })?;
+
+        let step = promql_parser::util::parse_duration(&query.step)
+            .map_err(|msg| BoxedError::new(PlainError::new(msg, StatusCode::InvalidArguments)))
+            .context(QueryParseSnafu {
+                query: &query.query,
+            })?;
 
         let eval_stmt = EvalStmt {
-            expr: prom_expr,
-            start: std::time::UNIX_EPOCH,
-            end: std::time::UNIX_EPOCH
-                .checked_add(Duration::from_secs(100))
-                .unwrap(),
-            interval: Duration::from_secs(5),
-            lookback_delta: Duration::from_secs(1),
+            expr,
+            start,
+            end,
+            interval: step,
+            // TODO(ruihang): provide a way to adjust this parameter.
+            lookback_delta: Duration::from_secs(DEFAULT_LOOKBACK),
         };
 
         Ok(QueryStatement::Promql(eval_stmt))
     }
+
+    fn parse_promql_timestamp(timestamp: &str) -> Result<SystemTime> {
+        // try rfc3339 format
+        let rfc3339_result = DateTime::parse_from_rfc3339(timestamp)
+            .context(ParseTimestampSnafu { raw: timestamp })
+            .map(Into::<SystemTime>::into);
+
+        // try float format
+        let float_result = timestamp
+            .parse::<f64>()
+            .context(ParseFloatSnafu { raw: timestamp })
+            .map(|float| {
+                let duration = Duration::from_secs_f64(float);
+                SystemTime::UNIX_EPOCH
+                    .checked_add(duration)
+                    .unwrap_or(max_system_timestamp())
+            });
```

Comment:
we can refine this flow to avoid parsing timestamp twice for all input. If it matches rfc3339 we can return early. 

also perhaps we can avoid this `NeitherError` by boxing rfc3339 error only, since the error for parsing float may not carry lots of information.

Reply:
I add shorthand logic in [c0c179c](https://github.com/GreptimeTeam/greptimedb/pull/985/commits/c0c179cb485bbd52ca3e5840a75e25c7ffe5f534) 😋

>also perhaps we can avoid this NeitherError by boxing rfc3339 error only, since the error for parsing float may not carry lots of information.

Do you mean always return rfc3339 error and discard the float parsing error?

Reply:
Yes. We can describe float parsing error in message and wrap rfc3339 error only. It's not a big deal, just to simplify the error object only.

Reply:
Makes sense, I change this in [c33e078](https://github.com/GreptimeTeam/greptimedb/pull/985/commits/c33e0788ecb34a615eec07dd73e98188fea619ab) and also remove `CompoundError` in [67c1804](https://github.com/GreptimeTeam/greptimedb/pull/985/commits/67c1804e43d34943b2952950455f9fc9bb5598b6)